### DISCUSSION TITLE
Slow simulation for RAM

### DISCUSSION THREAD

**Discussion by VerdichteteEnergie** [hneemann/Digital#1102] [General]
First of all, thank you for this great program. I love it. Before that I used Logisim, but it often causes problems with memory from custom logic gates. On the other hand, the memory works very well with the hneemann simulator.

First I built a 4-bit and 8-bit CPU with hneemann with 4-bit x16 and 8-bit x16 RAM . That worked well. Now I got it working with 16-bit x64 RAM. And it actually still works, but the simulation is very very slow.

Hence my question: Is there a way to give the simulation more computing power from my real computer? Or is it even possible that my computer (the real machine, not the simulation) is too slow for such a simulation?

It's an AMD Ryzen 3 1200 Quad and 16GB of RAM. It shouldn't be because of that, should it?

I checked the task manager and saw that "digital" is using about 30% CPU power and only 1GB of RAM.

I tried running the program with
`java -jar -Xmx2048M Digital.jar`
as suggested [here](https://github.com/hneemann/Digital/issues/278#issuecomment-496887657) by hneemann in another thread.
Unfortunately it didn't bring any change.

**Comment by hneemann** [comment]
It's hard to tell without seeing the circuit. Can you zip the circuit and attach it here?
(Open the main circuit and click _File->Export->Export as ZIP_)

**Reply by VerdichteteEnergie** [reply]
Cool that you answer personally!

I hope I'm not wasting your time. I'm just an amateur and interested in computer engineering and informatics.

I modeled the gates and circuits based on the [nand2tetris](https://www.nand2tetris.org/) course from the University of Jerusalem.

Here is my RAM curcuit [RAM64 (16-Bit).zip](https://github.com/hneemann/Digital/files/10425489/RAM64.16-Bit.zip)

**Reply by hneemann** [reply]
You have built the memory completely from transistors. Although this is pretty cool, your implementation needs about 80000 components for only 1024 bit s-ram. With such a large number of components, the simulator becomes slow.

You can reduce the consumption significantly by removing all pull-down resistors. They are not needed. You can also build an inverter with only one transistor and a pull-up resistor. This is more efficient than using a NAND.

But if you would build the memory like it is built in reality, you would need only a little bit more than 10000 components. (6 transistors and 4 resistors per bit + some control logic).

Have a look at the _cmos/sram.dig_ example.

**Reply by VerdichteteEnergie** [reply]
Thank You very much!

After removing the pull-down resistor in the NAND gate it runs very fast.

I will adjust the NOT gate as you said to get even better performance.

If I build the computer in real life, I want to optimize it.

I'm trying to understand how to get from logic gates to software. Because I'm learning to program and I'm interested in how a computer works at the hardware level and can then be controlled via software.

And thanks to you, besides HDL, now I can also grab it visually.

Your program is great!

**Reply by hneemann** [reply]
@VerdichteteEnergie 
This is a 64x64 RAM build with viewer components.
I'll leave it to you to replace the multiplexers, drivers and inverters with transitor based components. :smile:

[RAM64_64.zip](https://github.com/hneemann/Digital/files/10427335/RAM64_64.zip)

**Comment by VerdichteteEnergie** [comment]
I apologize for the inconvenience. But I have another question.

16-bit RAM memory is used in the nand2tetris course. The largest RAM bar has a 16K capacity 😟. Is it somehow possible to simulate this  memory in a stable/fast version in Digital? The bars have a total of 14 selectors for the addresses.

Is it possible that I can use the 16-bit sram from the Digital examples for this?

**Reply by hneemann** [reply]
That would require more than a million transistors. 
This cannot be realized with transistors only, in a way that the circuit remains usable. 
In this case it is better to use the built-in RAM component.

**Reply by VerdichteteEnergie** [reply]
Thank You for answer!

I probably worded my question wrong. I don't have curcuit yet. Because the performance with the 512 RAM was already so weak on my computer.

It's still the same principle as the zip file I posted earlier.

I'm just about having a 16K RAM 16-bit bar at all. It doesn't have to be the nand2tetris variant. Or do you or someone else know if it has already managed to simulate the nand2tetris?

So my question is: How can I build a 16K RAM stick in Digital with the given possibilities of the program itself.

**Reply by hneemann** [reply]
You don't need to build one. 
Just click **Components->Memory->RAM->RAM, dual Port** and place the RAM where you like it.

**Comment by VerdichteteEnergie** [comment]
After assembling the gates and chips myself and understanding them, I have now exchanged the gates for gates from the digital software.

I replaced all RAM parts. The CPU registers are an exception, because errors occurred after I tried to use the registers from digital.

As a result, the computer has become very fast. But it's not quite enough.

Maybe someone here can see how I can exchange some chips with parts from the digital software to improve performance?

Here is my version of the nand2tetris Hack-Computer: 
[hack-com.zip](https://github.com/hneemann/Digital/files/11121827/hack-com.zip)

**Reply by hneemann** [reply]
My first impression is that far too many splitters/mergers are used. If possible, you should try to stay with the buses. Splitting a bus and then merging it again is very inefficient. Also muxer/demuxer e.g. can switch complete buses.

**Reply by VerdichteteEnergie** [reply]
Many thanks for your response! 

Is it possible to have inputs and outputs for multiple bits on just one bus?

So I can use them as an output in a higher level chip?

**Reply by hneemann** [reply]
Yes! Inputs and outputs can also have multiple bits and thus connect multiple bits at the same time.

**Reply by prlaba** [reply]
Helmut, is there a performance penalty for using subcircuits?  If I have a main circuit and several subcircuits, will it perform slower than if I had everything in a single circuit?

Any other suggestions for improving a circuit's performance?

Thanks.

**Reply by hneemann** [reply]
@prlaba 
No, the use of subcircuits does not affect the performance of the simulation at all.
This is because the entire circuit is converted to a "flat" model when the simulation is started. All subcircuits are connected directly to the wires of the embedding circuit and all inputs and outputs are removed. If a subcircuit is used multiple times, the same number of copies will be inserted into the model. 
Only the start itself takes a few ms longer, but when the simulation is running, there is no difference anymore.
In fact, in both cases exactly the same simulation model is generated, no matter if subcircuits are used or not.

> Any other suggestions for improving a circuit's performance?

What should be avoided are relays and FET's. The simulation of these components is comparatively slow. 
What could also bring some performance is the avoidance of high-z busses, to which many outputs are connected. But this is not a very big effect. It will be noticeable only in rare cases.

**Reply by VerdichteteEnergie** [reply]
I have now replaced everything and replaced with chips and gates from the digital software.
The only thing that needs to be changed are the registers.
I was able to replace the register in the program counter with the register from digital software without any problems.
Unfortunately, the two CPU registers could not be exchanged because of problems with the ouput.

I built the registers from the the nand2tetris course and they work.
But as soon as I exchange them with the registers of the digital software, they no longer work...

Do you know how I can change my registers so that I can use the registers from the digital software?
[hack-reg.zip](https://github.com/hneemann/Digital/files/11131301/hack-reg.zip)

**Reply by hneemann** [reply]
@VerdichteteEnergie 
The handmade master-slave flip-flip you use uses a falling edge to propagate the input bit. 
Have you tried inverting the clock input of the built-in flip-flops to reproduce this behavior?

**Reply by VerdichteteEnergie** [reply]
It works with the inversion!

Thank you very much!
